File: howl-application/src/main/java/io/quartic/howl/HowlApplication.java
package io.quartic.howl;

import io.dropwizard.setup.Bootstrap;
import io.dropwizard.setup.Environment;
import io.dropwizard.websockets.WebsocketBundle;
import io.quartic.common.application.ApplicationBase;
import io.quartic.howl.storage.ObservableStorageBackend;
import io.quartic.howl.storage.StorageBackend;

import javax.websocket.server.ServerEndpointConfig;

import static io.quartic.common.websocket.WebsocketUtilsKt.serverEndpointConfig;

public class HowlApplication extends ApplicationBase<HowlConfiguration> {
    private WebsocketBundle websocketBundle = new WebsocketBundle(new ServerEndpointConfig[]{});

    public static void main(String[] args) throws Exception {
        new HowlApplication().run(args);
    }

    @Override
    public void initializeApplication(Bootstrap<HowlConfiguration> bootstrap) {
        bootstrap.addBundle(websocketBundle);
    }

    @Override
    public void runApplication(HowlConfiguration configuration, Environment environment) {
        StorageBackend storageBackend = configuration.getStorage().build();
        ObservableStorageBackend observableStorageBackend = new ObservableStorageBackend(storageBackend);
        environment.jersey().register(new HowlResource(observableStorageBackend));
        websocketBundle.addEndpoint(serverEndpointConfig("/changes/{namespace}/{objectName}",
                new WebsocketEndpoint(observableStorageBackend.changes())));
    }
}

File: howl-application/src/main/kotlin/io/quartic/howl/HowlApplication.kt
package io.quartic.howl

import io.dropwizard.setup.Bootstrap
import io.dropwizard.setup.Environment
import io.dropwizard.websockets.WebsocketBundle
import io.quartic.common.application.ApplicationBase
import io.quartic.common.websocket.serverEndpointConfig
import io.quartic.howl.storage.ObservableStorage
import io.quartic.howl.storage.RoutingStorage
import javax.websocket.server.ServerEndpointConfig

class HowlApplication : ApplicationBase<HowlConfiguration>() {
    private val websocketBundle = WebsocketBundle(*arrayOf<ServerEndpointConfig>())

    public override fun initializeApplication(bootstrap: Bootstrap<HowlConfiguration>) {
        bootstrap.addBundle(websocketBundle)
    }

    public override fun runApplication(configuration: HowlConfiguration, environment: Environment) {
        val storage = ObservableStorage(RoutingStorage(configuration.namespaces))
        environment.jersey().register(HowlResource(storage))
        websocketBundle.addEndpoint(serverEndpointConfig(
                "/changes/{namespace}/{objectName}",
                WebsocketEndpoint(storage.changes)
        ))
    }

    companion object {
        @JvmStatic fun main(args: Array<String>) = HowlApplication().run(*args)
    }
}
